page_group.php

```diff
@@ -0,0 +1,145 @@
+
+admin_externalpage_setup('tool_excimer_report_page_groups');
+
+$output = $PAGE->get_renderer('tool_excimer');
+
+$data = $pagegroup->to_record();
+
+// Data for table.
+$data->month = userdate(strtotime($data->month . '01'), '%b %Y');
+$data->approxcount = pow(2, $data->fuzzycount - 1) . ' - ' . pow(2, $data->fuzzycount);
+$data->approxduration = pow(2, $data->fuzzydurationsum);
+$data->histogram = helper::make_histogram($data);
+
+// Data for charts.
+// Each duration range forms its own line on the chart.
+
+$firstmonth = (int) userdate(strtotime("$monthstodisplay months ago"), '%Y%m'); // YYYYMM format.
+
+$history = $DB->get_records_select(
+    page_group::TABLE,
+    "name = ? and month >= ?",
+    [$data->name, $firstmonth],
+    '',
+    'month, id, fuzzydurationcounts'
+);
+
+$highest = 0;
+
+// Each of these is an array holding counts for each month for a particular duration range.
+$durationseries = [];
+
+$histograms = [];
+// We make a list of histograms, while finding the highest duration range that a histogram has a count for.
+foreach ($history as $month => $record) {
+    $histograms[$month] = helper::make_histogram($record);
+    $highest = max($highest, count($histograms[$month]));
+}
+
+// Labels for the axis.
+$labels = [];
+
+// Label for each line.
+$linelabels = [];
+
+$month = $firstmonth;
+// We count through the months to get them in the right order, and to capture months that have no record in the database.
+for ($i = 0; $i < $monthstodisplay; ++$i) {
+    $labels[] = userdate(strtotime($month . '01'), '%b %Y');
+    // For each histogram, the duration counts are stored in a simple array. We count from 0 to highest here to
+    // ensure that we have the same number of entries for each month.
+    for ($rangehigh = 0; $rangehigh < $highest; ++$rangehigh) {
+        if (isset($histograms[$month][$rangehigh])) {
+            $durationseries[$rangehigh][] = $histograms[$month][$rangehigh]['value'];
+            $linelabels[$rangehigh] = get_string('fuzzydurationrange', 'tool_excimer', $histograms[$month][$rangehigh]);
+        } else {
+            $durationseries[$rangehigh][] = 0;
+        }
+    }
+    // Increment the month, turning over the year if we get past December.
+    ++$month;
+    if ($month % 100 == 13) {
+        $month += 88;
+    }
+}
+
+$chart = new chart_line();
+$chart->set_labels($labels);
+
+// Add each duration range to the chart, but only those that have non-zero counts.
+foreach ($durationseries as $idx => $series) {
+    if (max($series) != 0) {
+        $chart->add_series(new chart_series($linelabels[$idx], $series));
+    }
+}
+
+$monthstodisplayurl = clone $url;
+$monthstodisplayurl->params(['id' => $pagegroupid]);
+$monthstodisplayselect = new \single_select(
+    $monthstodisplayurl,
+    'monthstodisplay',
+    array_slice(range(0, $expiry), 1, null, true),
+    $monthstodisplay
+);
+$monthstodisplayselect->set_label(get_string('months_to_display', 'tool_excimer'));
```

Comment:
A comment for this block could be useful here. Took me a while to figure out it was a combo-box with a list of months (back) to display.

Reply:
done